### ISSUE TITLE
Redundent lock for network_security_rule

### ISSUE THREAD

**Issue by magodo** [terraform-providers/terraform-provider-azurerm#5119]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version

Terraform v0.12.16

v1.38.0

### Affected Resource(s)

* `azurerm_arm_network_security_rule`

### Expected Behavior

The lock is not needed for CRUD in this resource, the concurrent race should be handled on service side, not client side.

### Actual Behavior

The locks here will tramendously slow down performance when user tries to add a big amount of NSRs into NSG.

### Detail

IMHO, the lock is necessary in cases when service do not have the concrete resource available, while terraform has the need to split something out from a resource.  For example the `resource_arm_iothub_route` to `resource_arm_iothub`, since service didn't provide direct rest api to iothub route.

However, in case of network security rule, there exists [rest api](https://docs.microsoft.com/en-us/rest/api/virtualnetwork/securityrules/createorupdate) to manage the rule directly. Then the lock against NSG is not needed here.

(PS: If this is not a false positive issue, then the similar situation exists in other resources also.)

**Comment by magodo** [comment]
I scan through all resources and found following contains lock, i think we'd better to identify one by one for the necessity of lock:

- resource_arm_iothub_fallback_route
- resource_arm_notification_hub_authorization_rule
- resource_arm_iothub_endpoint_servicebus_topic
- resource_arm_network_interface_nat_rule_association
- resource_arm_firewall_nat_rule_collection
- resource_arm_iothub_shared_access_policy
- resource_arm_firewall_application_rule_collection
- resource_arm_network_interface_application_security_group_association
- resource_arm_network_interface_application_gateway_association
- resource_arm_network_profile
- resource_arm_iothub
- resource_arm_iothub_endpoint_storage_container
- resource_arm_firewall
- resource_arm_key_vault
- resource_arm_subnet
- resource_arm_iothub_route
- resource_arm_logic_app_workflow
- resource_arm_virtual_hub
- resource_arm_iothub_endpoint_eventhub
- resource_arm_virtual_machine_data_disk_attachment
- resource_arm_route
- resource_arm_ddos_protection_plan
- resource_arm_subnet_route_table_association
- resource_arm_app_service_source_control_token
- resource_arm_key_vault_access_policy
- resource_arm_subnet_nat_gateway_association
- resource_arm_iothub_endpoint_servicebus_queue
- resource_arm_virtual_machine
- resource_arm_firewall_network_rule_collection
- resource_arm_redis_cache
- resource_arm_network_security_rule
- resource_arm_network_ddos_protection_plan
- resource_arm_network_interface
- resource_arm_storage_account_network_rules
- resource_arm_express_route_circuit
- resource_arm_azuread_service_principal_password
- resource_arm_express_route_circuit_authorization
- resource_arm_network_security_group
- resource_arm_network_interface_backend_address_pool_association
- resource_arm_express_route_circuit_peering
- resource_arm_storage_account
- resource_arm_subnet_network_security_group_association
- resource_arm_app_service_custom_hostname_binding

**Comment by tombuildsstuff** [comment]
hi @magodo 

Thanks for opening this issue.

We use locks within the Azure Provider when we've identified bugs in the API which require this (for non-virtual-resources) - as such in this case I recall this coming from a user seeing an API response of (paraphrasing) "cannot make concurrent changes" which is why this exists.

Whilst we've tried raising these bugs with the Service Teams - chatting with the the Networking team in the past their opinion is that it's on the Client Application to handle this rather than being a design bug in their service. As such whilst this isn't ideal from a performance perspective - we'd rather ship something that took a little longer to provision but worked, than something which ultimately didn't - but since this is by design to work around bugs in the Azure API I'm going to close this issue for the moment.

Thanks!
